File: scripts/prepare-submission.js
import { mkdir, writeFile } from "node:fs/promises";
import { DEMO_TEXT } from "../src/app.js";
import { finish, validateExtraction } from "../src/normalize.js";
await mkdir("docs", { recursive: true });
await mkdir("submission", { recursive: true });
const extraction = {
  tests_raw: DEMO_TEXT.split("\n").slice(1),
  uncertain: false,
};
const example = finish(
  validateExtraction(DEMO_TEXT, extraction),
  "demo",
  DEMO_TEXT,
  extraction,
);
const writeJSON = (path, value) =>
  writeFile(path, JSON.stringify(value, null, 2) + "\n");
await writeJSON("samples/expected-response.json", example);
const errorSchema = {
  type: "object",
  required: ["status", "reason"],
  properties: {
    status: { enum: ["error", "unprocessed"] },
    reason: { type: "string" },
  },
};
const successSchema = {
  type: "object",
  required: ["tests", "summary", "status", "explanations", "metadata"],
  properties: {
    tests: {
      type: "array",
      minItems: 1,
      items: {
        type: "object",
        required: ["name", "value", "unit", "status", "ref_range"],
        properties: {
          name: { type: "string" },
          value: { type: "number" },
          unit: { type: "string" },
          status: { enum: ["low", "high", "normal", "unknown"] },
          ref_range: {
            anyOf: [
              { type: "null" },
              {
                type: "object",
                required: ["low", "high"],
                properties: {
                  low: { type: "number" },
                  high: { type: "number" },
                },
              },
            ],
          },
        },
      },
    },
    summary: { type: "string" },
    status: { const: "ok" },
    explanations: {
      type: "array",
      items: {
        type: "object",
        required: ["name", "text"],
        properties: { name: { type: "string" }, text: { type: "string" } },
      },
    },
    metadata: {
      type: "object",
      required: [
        "mode",
        "source_text",
        "tests_raw",
        "evidence",
        "validation",
        "confidence",
        "confidence_note",
        "note",
      ],
      properties: {
        mode: { enum: ["demo", "text", "image"] },
        source_text: { type: "string" },
        tests_raw: { type: "array", items: { type: "string" } },
        evidence: {
          type: "array",
          items: {
            type: "object",
            required: ["name", "source"],
            properties: {
              name: { type: "string" },
              source: { type: "string" },
            },
          },
        },
        validation: { type: "string" },
        confidence: { type: "null" },
        confidence_note: { type: "string" },
        note: { type: "string" },
      },
    },
  },
};
const responses = {
  200: {
    description: "Validated report",
    content: {
      "application/json": {
        schema: { $ref: "#/components/schemas/Report" },
        example,
      },
    },
  },
};
for (const code of [400, 413, 422, 429, 500])
  responses[code] = {
    description: {
      400: "Invalid input",
      401: "Missing key",
      413: "Body too large",
      422: "Unprocessed report",
      429: "Busy or quota exceeded",
      502: "Provider failure",
      504: "Provider timeout",
      500: "Unexpected server failure",
    }[code],
    content: {
      "application/json": { schema: { $ref: "#/components/schemas/Error" } },
    },
  };
await writeJSON("docs/openapi.json", {
  openapi: "3.1.0",
  info: {
    title: "Plum Report Simplifier",
    version: "1.0.0",
    description:
      "Deterministic text processing and local OCR. Optional per-request Gemini key for constrained wording. Provider failures use explicit template fallback.",
  },
  servers: [{ url: "http://127.0.0.1:3000" }],
  components: {
    securitySchemes: {
      GeminiKey: { type: "apiKey", in: "header", name: "x-gemini-api-key" },
    },
    schemas: { Report: successSchema, Error: errorSchema },
  },
  paths: {
    "/health": {
      get: {
        summary: "Health and optional server-key availability",
        responses: { 200: { description: "Service is running" } },
      },
    },
    "/api/demo": {
      get: {
        summary: "Fixed demo fixture; no API key or provider calls",
        responses: { 200: responses["200"] },
      },
    },
    "/reports/simplify/text": {
      post: {
        summary: "Simplify typed report",
        security: [{}, { GeminiKey: [] }],
        requestBody: {
          required: true,
          content: {
            "application/json": {
              schema: {
                type: "object",
                required: ["text"],
                additionalProperties: false,
                properties: {
                  text: { type: "string", minLength: 1, maxLength: 20000 },
                },
              },
              example: { text: DEMO_TEXT },
            },
          },
        },
        responses,
      },
    },
    "/reports/simplify/image": {
      post: {
        summary:
          "Local OCR and simplify one PNG, JPEG or WebP, up to 5 MB and 20 megapixels",
        security: [{ GeminiKey: [] }],
        requestBody: {
          required: true,
          content: {
            "multipart/form-data": {
              schema: {
                type: "object",
                required: ["image"],
                properties: { image: { type: "string", format: "binary" } },
              },
            },
          },
        },
        responses,
      },
    },
  },
});
function item(
  name,
  path,
  { method = "GET", body, status = 200, auth = false } = {},
) {
  return {
    name,
    request: {
      method,
      header: auth
        ? [{ key: "x-gemini-api-key", value: "{{gemini_api_key}}" }]
        : [],
      url: "{{base_url}}" + path,
      ...(body ? { body } : {}),
      description: auth
        ? "Set gemini_api_key locally in Postman. Never export a populated credential."
        : "No credentials needed.",
    },
    event: [
      {
        listen: "test",
        script: {
          type: "text/javascript",
          exec: [
            `pm.test('Expected HTTP status', () => pm.response.to.have.status(${status}));`,
            "pm.test('JSON response', () => pm.expect(pm.response.json()).to.have.property('status'));",
            ...(status === 200 && path !== "/health"
              ? [
                  "pm.test('Two sample results', () => pm.expect(pm.response.json().tests).to.have.lengthOf(2));",
                ]
              : []),
          ],
        },
      },
    ],
  };
}
const textBody = (text) => ({
  mode: "raw",
  raw: JSON.stringify({ text }, null, 2),
  options: { raw: { language: "json" } },
});
await writeJSON("docs/Plum.postman_collection.json", {
  info: {
    name: "Plum submission API",
    schema:
      "https://schema.getpostman.com/json/collection/v2.1.0/collection.json",
  },
  variable: [
    { key: "base_url", value: "http://127.0.0.1:3000" },
    { key: "gemini_api_key", value: "", type: "string" },
  ],
  item: [
    item("Health", "/health"),
    item("Demo without a key", "/api/demo"),
    item("Live text report", "/reports/simplify/text", {
      method: "POST",
      auth: true,
      body: textBody(DEMO_TEXT),
    }),
    item("Live image report", "/reports/simplify/image", {
      method: "POST",
      auth: true,
      body: {
        mode: "formdata",
        formdata: [
          { key: "image", type: "file", src: "../samples/report.png" },
        ],
      },
    }),
    item("Reject conflicting flag", "/reports/simplify/text", {
      method: "POST",
      auth: true,
      status: 422,
      body: textBody("Hemoglobin 10.2 g/dL (High) Reference: 12-15"),
    }),
    item("Core processing without key", "/reports/simplify/text", {
      method: "POST",
      status: 200,
      body: textBody(DEMO_TEXT),
    }),
  ],
});
console.log(
  "Prepared OpenAPI spec, Postman collection, and expected response fixture.",
);

File: scripts/prepare_submission.py
"""Generate reproducible sample inputs and submission artifacts."""

import asyncio
import json
from pathlib import Path

from PIL import Image, ImageDraw, ImageFont

from app.main import app
from app.pipeline import ASSIGNMENT_DEMO, ASSIGNMENT_INPUT

ROOT = Path(__file__).resolve().parents[1]


def save_json(path: str, value):
    target = ROOT / path
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(json.dumps(value, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")


def image_fixture():
    fonts = ["C:/Windows/Fonts/arial.ttf", "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf"]
    font = next(
        (ImageFont.truetype(p, 32) for p in fonts if Path(p).exists()),
        ImageFont.load_default(size=32),
    )
    image = Image.new("RGB", (1400, 360), "white")
    draw = ImageDraw.Draw(image)
    for index, line in enumerate(ASSIGNMENT_DEMO.splitlines()):
        draw.text((45, 35 + index * 95), line, fill="black", font=font)
    image.save(ROOT / "samples/report.png")


async def main():
    (ROOT / "samples").mkdir(exist_ok=True)
    fixtures = {
        "report.txt": ASSIGNMENT_DEMO,
        "report-original.txt": ASSIGNMENT_INPUT,
        "report-ocr-typos.txt": "CBC: Hemglobin 10.2 g/dL (Low)\nWBC 11200 /uL (Hgh)",
        "report-conflict.txt": "Hemoglobin 10.2 g/dL (High) Reference: 12-15",
        "report-units.txt": "Hemoglobin 102 g/L (Low) Reference: 120-150\nWBC 11.2 10^9/L (High) Reference: 4-11",
    }
    for filename, text in fixtures.items():
        (ROOT / "samples" / filename).write_text(text + "\n", encoding="utf-8")
    image_fixture()
    example = await app.state.pipeline.run(ASSIGNMENT_DEMO, "sample-request", "assignment_demo")
    example.metadata.duration_ms = 0
    save_json("samples/expected-response.json", example.model_dump())
    save_json("docs/openapi.json", app.openapi())
    items = []

    def add(name, method, endpoint, body=None, expected=200, key=False):
        request = {"method": method, "url": "{{base_url}}" + endpoint, "header": []}
        if key:
            request["header"].append({"key": "X-Gemini-API-Key", "value": "{{gemini_api_key}}"})
        if body is not None:
            request["body"] = body
        items.append(
            {
                "name": name,
                "request": request,
                "event": [
                    {
                        "listen": "test",
                        "script": {
                            "type": "text/javascript",
                            "exec": [
                                f"pm.test('Expected HTTP status', () => pm.response.to.have.status({expected}));",
                                "pm.test('JSON response', () => pm.expect(pm.response.json()).to.have.property('status'));",
                            ],
                        },
                    }
                ],
            }
        )

    def raw(text, use_gemini=False):
        return {
            "mode": "raw",
            "raw": json.dumps({"text": text, "use_gemini": use_gemini}),
            "options": {"raw": {"language": "json"}},
        }

    add("Health", "GET", "/health")
    add("Assignment demo", "GET", "/demo")
    add("Text without LLM", "POST", "/reports/simplify/text", raw(ASSIGNMENT_DEMO))
    add(
        "Text with Gemini language",
        "POST",
        "/reports/simplify/text",
        raw(ASSIGNMENT_DEMO, True),
        key=True,
    )
    add(
        "Local image OCR",
        "POST",
        "/reports/simplify/image",
        {
            "mode": "formdata",
            "formdata": [{"key": "image", "type": "file", "src": "../samples/report.png"}],
        },
    )
    add(
        "Conflicting flag is rejected",
        "POST",
        "/reports/simplify/text",
        raw(fixtures["report-conflict.txt"]),
        expected=422,
    )
    add(
        "Gemini key required only when enabled",
        "POST",
        "/reports/simplify/text",
        raw(ASSIGNMENT_DEMO, True),
        expected=401,
    )
    save_json(
        "docs/Plum.postman_collection.json",
        {
            "info": {
                "name": "Medical Report Reader",
                "schema": "https://schema.getpostman.com/json/collection/v2.1.0/collection.json",
            },
            "variable": [
                {"key": "base_url", "value": "http://127.0.0.1:8000"},
                {"key": "gemini_api_key", "value": ""},
            ],
            "item": items,
        },
    )
    print("Generated synthetic fixtures, expected response, OpenAPI and Postman collection.")


if __name__ == "__main__":
    asyncio.run(main())
